source/game/system/map/MapdataCheckPoint.cc

```diff
@@ -18,17 +21,110 @@ MapdataCheckPoint::MapdataCheckPoint(const SData *data)
-    m_nextPt = stream.read_u8();
+    m_prevPt = stream.read_s8();
+    m_nextPt = stream.read_s8();
+}
+
+/// @brief Calculates @ref m_nextPoints and @ref m_prevPoints from @ref m_nextPt and @ref m_prevPt.
+/// @details Also calculates the quadrilaterals for the next checkpoints, filling the fields of @ref
+/// LinkedCheckpoint for each.
+/// @addr{0x80515624}
+void MapdataCheckPoint::initCheckpointLinks(MapdataCheckPointAccessor &accessor, int id) {
+    m_id = id;
+    auto checkPathAccessor = CourseMap::Instance()->getCheckPathAccessor();
+
+    // Calculate the quadrilateral's `m_prevPoints`. If the check point is the first in its group,
+    // it has multiple previous checkpoints defined by its preceding checkpaths
+    if (m_prevPt == -1) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        m_prevCount = 0;
+
+        for (auto [i, prevID] : std::views::enumerate(checkpath->prev())) {
+            if (prevID == -1) {
+                continue;
+            }
+
+            MapdataCheckPath *prev = checkPathAccessor->get(prevID);
+            m_prevPoints[i] = accessor.get(prev->end());
+            m_prevCount++;
+        }
+    } else {
+        m_prevPoints[0] = accessor.get(m_prevPt);
+        m_prevCount++;
+    }
+
+    // Calculate the quadrilateral's `m_nextPoints`. If the checkpoint is the last in its group, it
+    // can have multiple quadrilaterals (and nextCheckpoint) which are determined by its next
+    // path(s)
+    if (m_nextPt == -1) {
+        MapdataCheckPath *checkpath = checkPathAccessor->findCheckpathForCheckpoint(id);
+        m_nextCount = 0;
+
+        for (auto [i, nextID] : std::views::enumerate(checkpath->next())) {
+            if (nextID == -1) {
+                continue;
+            }
+
+            MapdataCheckPath *next = checkPathAccessor->get(nextID);
+            m_nextPoints[i].checkpoint = accessor.get(next->start());
+            m_nextCount++;
+        }
+    } else {
+        m_nextPoints[0].checkpoint = accessor.get(m_nextPt);
+        m_nextCount++;
+    }
+
+    // Form the checkpoint's quadrilateral(s)
+    for (size_t i = 0; i < m_nextCount; i++) {
+        auto &nextLinked = m_nextPoints[i];
+        auto *nextPoint = nextLinked.checkpoint;
+
+        nextLinked.distance = (nextPoint->m_midpoint - m_midpoint).normalise();
+        nextLinked.p0diff = nextPoint->m_left - m_left;
+        nextLinked.p1diff = nextPoint->m_right - m_right;
+    }
+}
+
+s8 MapdataCheckPoint::checkArea() const {
+    return m_checkArea;
+}
+
+MapdataCheckPoint *MapdataCheckPoint::nextPoint(size_t i) const {
+    ASSERT(i < m_nextPoints.size());
+    return m_nextPoints[i].checkpoint;
+}
+
+/// @brief Finds indices of the finish line and last key checkpoint; also calls initCheckpointLinks
+/// for all checkpoints
+/// @addr{0x80515244}
+void MapdataCheckPointAccessor::init() {
+    s8 lastKcpType = -1;
+    s16 finishLineCheckpointId = -1;
```

Comment:
This is probably a question for vabold - we could improve this function by getting rid of these stack variables and just directly assigning to the members in the loop

Reply:
sounds good to me, but ill wait for @vabold to confirm

Reply:
The function is more optimized with stack variables (due to the compiler being allowed to optimize into non-volatile GPRs) than constantly writing to RAM.